src/create_venv.sh

```diff
@@ -1,136 +1,258 @@
+    # correct requirements.txt and frozen_requirements.txt
+    if [[ "${unique_sha}" == "$(cat "${venv_dir}/YB_VENV_SHA")" ]]; then
+      verbose "Existing venv is current and will be used as is."
+      return 1
+    fi
+  fi
+  verbose "The venv needs refreshing"
+  true
+}
 
-echo "Using root_dir=${root_dir}"
-echo "Using reqs_file=${reqs_file}"
+# -------------------------------------------------------------------------------------------------
+# Main functions
+# -------------------------------------------------------------------------------------------------
+function yb_deactivate_virtualenv() {
+  # shellcheck disable=SC2031
+  if [[ -n ${VIRTUAL_ENV:-} && -f "$VIRTUAL_ENV/bin/activate" ]]; then
+    set +u
+    # The "deactivate" function is defined by virtualenv's "activate" script.
+    deactivate
+    set -u
 
-# Include the OS and h/w arch.  This allows to use a VM or container with a shared persistent
-# externally mounted YB_VENV_BASE_DIR
-unique_input="$(uname -s)$(uname -m)${py_cmd}"
-if [[ ! -f "${reqs_file}" ]]; then
-  echo "WARNING: No requirements.txt file found!"
-  exit 1
-fi
-reqs_sha="$(text_file_sha "${reqs_file}")"
+    unset PYTHONPATH
+  fi
+}
 
-unique_input="${unique_input}$(sort -u "${reqs_file}")"
+# Arguments:
+#   - Parent directory of the virtualenv
+#   - Python interpreter to use (optional)
+function yb_activate_virtualenv() {
+  # Expand the path here, we don't want it to be .
+  local root_dir
+  root_dir=$(realpath "${1}")
+  local reqs_file="${root_dir}/requirements.txt"
+  local frozen_file="${root_dir}/requirements_frozen.txt"
 
-refreeze=false
-if [[ -f "${frzn_file}" ]]; then
-  unique_input="${unique_input}$(sort -u "${frzn_file}")"
-  if ! grep "# YB_SHA: ${reqs_sha}" "${frzn_file}" >/dev/null 2>&1; then
-    refreeze=true
+  # Allow the caller to optionally pass in a venv path to use instead of trying to calculate it
+  # This is to support the use case in yugabyte-db:yb_build.sh
+  local venv_dir=${2:-}
+  # By default we create a unique VENV dir based on a combination of python version, OS, arch,
+  # and the non-comment contents of the requirements.txt file.  If YB_PUT_VENV_IN_PROJECT_DIR is set
+  # true we fall back to the older behaviour of using a directory called 'venv' in the same
+  # directory that contains the requirements.txt file.  It is possible for this older style venv
+  # dir to go out of date in a way that is hard to detect.
+  # Include the OS and h/w arch.  This allows to use a VM or container with a shared persistent
+  # externally mounted YB_VENV_BASE_DIR
+  local unique_input
+  unique_input="$(uname -s)$(uname -m)${yb_python_version_actual}"
+  local refreeze=false
+  if [[ -f "${reqs_file}" ]]; then
+    local reqs_sha
+    reqs_sha="$(ybcv::text_file_sha_ignore_comments "${reqs_file}")"
+    unique_input="${unique_input}$(sort -u "${reqs_file}")"
+    if ybcv::needs_refreeze "${reqs_sha}" "${frozen_file}"; then
+      if ${YB_BUILD_STRICT}; then
```

Comment:
Check booleans like so: `[[ "${YB_BUILD_STRICT}" == "true" ]]` (not by executing the command stored in the variable).